File: listener.go
// Copyright 2025 OWASP CRS Project
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"fmt"
	"strconv"

	"github.com/antlr4-go/antlr/v4"

	"github.com/coreruleset/seclang_parser/parser"
)

type ParserResult struct {
	Comments              []string `yaml:"comments"`
	CommentsBlocks        int      `yaml:"comment_blocks"`
	Variables             []string `yaml:"variables"`
	NegatedVarCount       int      `yaml:"negated_var_count"`
	CollectionLengthCount int      `yaml:"collection_length_count"`
	Collections           []string `yaml:"collections"`
	CollectionArgs        []string `yaml:"collection_args"`
	OperatorList          []string `yaml:"operator_list"`
	OperatorValueList     []string `yaml:"operator_value_list"`
	NegatedOperatorCount  int      `yaml:"negated_operator_count"`
	DirectiveList         []string `yaml:"directive_list"`
	DirectiveValues       []string `yaml:"directive_values"`
	RangeEvents           []string `yaml:"range_events"`
	RangeStartEvents      []int    `yaml:"range_start_events"`
	RangeEndEvents        []int    `yaml:"range_end_events"`
	SetvarCollections     []string `yaml:"setvar_collections"`
	SetvarNames           []string `yaml:"setvar_names"`
	SetvarOperations      []string `yaml:"setvar_operations"`
}

type TreeShapeListener struct {
	*parser.BaseSecLangParserListener
	results ParserResult
}

func NewTreeShapeListener() *TreeShapeListener {
	return new(TreeShapeListener)
}

type CustomErrorListener struct {
	*antlr.DefaultErrorListener
	Errors []error
}

func NewCustomErrorListener() *CustomErrorListener {
	return &CustomErrorListener{antlr.NewDefaultErrorListener(), make([]error, 0)}
}

func (c *CustomErrorListener) SyntaxError(recognizer antlr.Recognizer, offendingSymbol interface{}, line, column int, msg string, e antlr.RecognitionException) {
	var err error
	if offendingSymbol == nil {
		err = fmt.Errorf("Recognition error at line %d, column %d: %s", line, column, msg)
	} else {
		err = fmt.Errorf("Syntax error at line %d, column %d: %v", line, column, offendingSymbol)
	}
	c.Errors = append(c.Errors, err)
}

func (t *TreeShapeListener) EnterEveryRule(ctx antlr.ParserRuleContext) {
	// if you need to debug, enable this one below
	// fmt.Println(ctx.GetText())
}

func (l *TreeShapeListener) EnterInt_range(ctx *parser.Int_rangeContext) {
	l.results.RangeEvents = append(l.results.RangeEvents, ctx.GetText())
}

func (l *TreeShapeListener) EnterRange_start(ctx *parser.Range_startContext) {
	i, err := strconv.Atoi(ctx.GetText())
	if err != nil {
		panic(err)
	}
	l.results.RangeStartEvents = append(l.results.RangeStartEvents, i)
}

func (l *TreeShapeListener) EnterRange_end(ctx *parser.Range_endContext) {
	i, err := strconv.Atoi(ctx.GetText())
	if err != nil {
		panic(err)
	}
	l.results.RangeEndEvents = append(l.results.RangeEndEvents, i)
}

func (l *TreeShapeListener) EnterOperator_int_range(ctx *parser.Operator_int_rangeContext) {
	l.results.RangeEvents = append(l.results.RangeEvents, ctx.GetText())
	i, err := strconv.Atoi(ctx.INT_RANGE_VALUE(0).GetText())
	if err != nil {
		panic(err)
	}
	l.results.RangeStartEvents = append(l.results.RangeStartEvents, i)
	i, err = strconv.Atoi(ctx.INT_RANGE_VALUE(1).GetText())
	if err != nil {
		panic(err)
	}
	l.results.RangeEndEvents = append(l.results.RangeEndEvents, i)
}

func (l *TreeShapeListener) EnterRemove_rule_by_id(ctx *parser.Remove_rule_by_idContext) {
	l.results.DirectiveList = append(l.results.DirectiveList, ctx.GetText())
}

func (l *TreeShapeListener) EnterRemove_rule_by_id_int(ctx *parser.Remove_rule_by_id_intContext) {
	l.results.DirectiveValues = append(l.results.DirectiveValues, ctx.GetText())
}

func (l *TreeShapeListener) EnterRemove_rule_by_id_int_range(ctx *parser.Remove_rule_by_id_int_rangeContext) {
	l.results.DirectiveValues = append(l.results.DirectiveValues, ctx.GetText())
}

func (l *TreeShapeListener) EnterRemove_rule_by_msg(ctx *parser.Remove_rule_by_msgContext) {
	l.results.DirectiveList = append(l.results.DirectiveList, ctx.GetText())
}

func (l *TreeShapeListener) EnterRemove_rule_by_tag(ctx *parser.Remove_rule_by_tagContext) {
	l.results.DirectiveList = append(l.results.DirectiveList, ctx.GetText())
}

func (l *TreeShapeListener) EnterString_remove_rules_values(ctx *parser.String_remove_rules_valuesContext) {
	l.results.DirectiveValues = append(l.results.DirectiveValues, ctx.GetText())
}

func (l *TreeShapeListener) EnterUpdate_target_by_id(ctx *parser.Update_target_by_idContext) {
	l.results.DirectiveList = append(l.results.DirectiveList, ctx.GetText())
}

func (l *TreeShapeListener) EnterUpdate_target_by_msg(ctx *parser.Update_target_by_msgContext) {
	l.results.DirectiveList = append(l.results.DirectiveList, ctx.GetText())
}

func (l *TreeShapeListener) EnterUpdate_target_by_tag(ctx *parser.Update_target_by_tagContext) {
	l.results.DirectiveList = append(l.results.DirectiveList, ctx.GetText())
}

func (l *TreeShapeListener) EnterUpdate_target_rules_values(ctx *parser.Update_target_rules_valuesContext) {
	l.results.DirectiveValues = append(l.results.DirectiveValues, ctx.GetText())
}

func (l *TreeShapeListener) EnterVariable_enum(ctx *parser.Variable_enumContext) {
	l.results.Variables = append(l.results.Variables, ctx.GetText())
}

func (l *TreeShapeListener) EnterCollection_enum(ctx *parser.Collection_enumContext) {
	l.results.Collections = append(l.results.Collections, ctx.GetText())
}

func (l *TreeShapeListener) EnterCollection_value(ctx *parser.Collection_valueContext) {
	l.results.CollectionArgs = append(l.results.CollectionArgs, ctx.GetText())
}

func (l *TreeShapeListener) EnterVar_not(ctx *parser.Var_notContext) {
	l.results.NegatedVarCount++
}

func (l *TreeShapeListener) EnterVar_count(ctx *parser.Var_countContext) {
	l.results.CollectionLengthCount++
}

func (l *TreeShapeListener) EnterEngine_config_rule_directive(ctx *parser.Engine_config_rule_directiveContext) {
	l.results.DirectiveList = append(l.results.DirectiveList, ctx.GetText())
}

func (l *TreeShapeListener) EnterConfig_dir_sec_action(ctx *parser.Config_dir_sec_actionContext) {
	l.results.DirectiveList = append(l.results.DirectiveList, ctx.GetText())
}

func (l *TreeShapeListener) EnterOperator_name(ctx *parser.Operator_nameContext) {
	l.results.OperatorList = append(l.results.OperatorList, ctx.GetText())
}

func (l *TreeShapeListener) EnterOperator_value(ctx *parser.Operator_valueContext) {
	l.results.OperatorValueList = append(l.results.OperatorValueList, ctx.GetText())
}

func (l *TreeShapeListener) EnterOperator_not(ctx *parser.Operator_notContext) {
	l.results.NegatedOperatorCount++
}

func (l *TreeShapeListener) EnterCol_name(ctx *parser.Col_nameContext) {
	l.results.SetvarCollections = append(l.results.SetvarCollections, ctx.GetText())
}

func (l *TreeShapeListener) EnterSetvar_stmt(ctx *parser.Setvar_stmtContext) {
	l.results.SetvarNames = append(l.results.SetvarNames, ctx.GetText())
}

func (l *TreeShapeListener) EnterAssignment(ctx *parser.AssignmentContext) {
	l.results.SetvarOperations = append(l.results.SetvarOperations, ctx.GetText())
}

func (l *TreeShapeListener) EnterVar_assignment(ctx *parser.Var_assignmentContext) {
	l.results.DirectiveValues = append(l.results.DirectiveValues, ctx.GetText())
}

func (l *TreeShapeListener) EnterCtl_action(ctx *parser.Ctl_actionContext) {
	l.results.DirectiveValues = append(l.results.DirectiveValues, ctx.GetText())
}

func (l *TreeShapeListener) EnterCtl_id(ctx *parser.Ctl_idContext) {
	l.results.DirectiveValues = append(l.results.DirectiveValues, ctx.GetText())
}

func (l *TreeShapeListener) EnterComment(ctx *parser.CommentContext) {
	// ctx.COMMENT() can be nil if there is only a HASH without comment text
	if ctx.COMMENT() != nil {
		l.results.Comments = append(l.results.Comments, ctx.COMMENT().GetText())
	} else {
		l.results.Comments = append(l.results.Comments, "")
	}
}

func (l *TreeShapeListener) EnterTransformation_action_value(ctx *parser.Transformation_action_valueContext) {
	l.results.DirectiveValues = append(l.results.DirectiveValues, ctx.GetText())
}

func (l *TreeShapeListener) EnterComment_block(ctx *parser.Comment_blockContext) {
	l.results.CommentsBlocks++
}

File: src/seclang_parser/listener.py
# Copyright 2025 OWASP CRS Project
# SPDX-License-Identifier: Apache-2.0

"""Test listener for collecting parser results."""

from dataclasses import dataclass, field

from seclang_parser.SecLangParserListener import SecLangParserListener
from seclang_parser.SecLangParser import SecLangParser


@dataclass
class ParserResult:
    """Stores results from parsing for test validation."""

    comments: list[str] = field(default_factory=list)
    comment_blocks: int = 0
    variables: list[str] = field(default_factory=list)
    negated_var_count: int = 0
    collection_length_count: int = 0
    collections: list[str] = field(default_factory=list)
    collection_args: list[str] = field(default_factory=list)
    operator_list: list[str] = field(default_factory=list)
    operator_value_list: list[str] = field(default_factory=list)
    negated_operator_count: int = 0
    directive_list: list[str] = field(default_factory=list)
    directive_values: list[str] = field(default_factory=list)
    range_events: list[str] = field(default_factory=list)
    range_start_events: list[int] = field(default_factory=list)
    range_end_events: list[int] = field(default_factory=list)
    setvar_collections: list[str] = field(default_factory=list)
    setvar_names: list[str] = field(default_factory=list)
    setvar_operations: list[str] = field(default_factory=list)


class TreeShapeListener(SecLangParserListener):
    """Listener that collects parsing results for test validation."""

    def __init__(self):
        super().__init__()
        self.results = ParserResult()

    def enterInt_range(self, ctx: SecLangParser.Int_rangeContext):
        self.results.range_events.append(ctx.getText())

    def enterRange_start(self, ctx: SecLangParser.Range_startContext):
        self.results.range_start_events.append(int(ctx.getText()))

    def enterRange_end(self, ctx: SecLangParser.Range_endContext):
        self.results.range_end_events.append(int(ctx.getText()))

    def enterRemove_rule_by_id(self, ctx: SecLangParser.Remove_rule_by_idContext):
        self.results.directive_list.append(ctx.getText())

    def enterRemove_rule_by_id_int(self, ctx: SecLangParser.Remove_rule_by_id_intContext):
        self.results.directive_values.append(ctx.getText())

    def enterRemove_rule_by_id_int_range(
        self, ctx: SecLangParser.Remove_rule_by_id_int_rangeContext
    ):
        self.results.directive_values.append(ctx.getText())

    def enterRemove_rule_by_msg(self, ctx: SecLangParser.Remove_rule_by_msgContext):
        self.results.directive_list.append(ctx.getText())

    def enterRemove_rule_by_tag(self, ctx: SecLangParser.Remove_rule_by_tagContext):
        self.results.directive_list.append(ctx.getText())

    def enterString_remove_rules_values(self, ctx: SecLangParser.String_remove_rules_valuesContext):
        self.results.directive_values.append(ctx.getText())

    def enterUpdate_target_by_id(self, ctx: SecLangParser.Update_target_by_idContext):
        self.results.directive_list.append(ctx.getText())

    def enterUpdate_target_by_msg(self, ctx: SecLangParser.Update_target_by_msgContext):
        self.results.directive_list.append(ctx.getText())

    def enterUpdate_target_by_tag(self, ctx: SecLangParser.Update_target_by_tagContext):
        self.results.directive_list.append(ctx.getText())

    def enterUpdate_target_rules_values(
        self, ctx: SecLangParser.Update_target_rules_valuesContext
    ):
        self.results.directive_values.append(ctx.getText())

    def enterVariable_enum(self, ctx: SecLangParser.Variable_enumContext):
        self.results.variables.append(ctx.getText())

    def enterCollection_enum(self, ctx: SecLangParser.Collection_enumContext):
        self.results.collections.append(ctx.getText())

    def enterCollection_value(self, ctx: SecLangParser.Collection_valueContext):
        self.results.collection_args.append(ctx.getText())

    def enterVar_not(self, ctx: SecLangParser.Var_notContext):
        self.results.negated_var_count += 1

    def enterVar_count(self, ctx: SecLangParser.Var_countContext):
        self.results.collection_length_count += 1

    def enterEngine_config_rule_directive(self, ctx: SecLangParser.Engine_config_rule_directiveContext):
        self.results.directive_list.append(ctx.getText())

    def enterOperator_name(self, ctx: SecLangParser.Operator_nameContext):
        self.results.operator_list.append(ctx.getText())

    def enterOperator_value(self, ctx: SecLangParser.Operator_valueContext):
        self.results.operator_value_list.append(ctx.getText())

    def enterOperator_not(self, ctx: SecLangParser.Operator_notContext):
        self.results.negated_operator_count += 1

    def enterCol_name(self, ctx: SecLangParser.Col_nameContext):
        self.results.setvar_collections.append(ctx.getText())

    def enterSetvar_stmt(self, ctx: SecLangParser.Setvar_stmtContext):
        self.results.setvar_names.append(ctx.getText())

    def enterAssignment(self, ctx: SecLangParser.AssignmentContext):
        self.results.setvar_operations.append(ctx.getText())

    def enterVar_assignment(self, ctx: SecLangParser.Var_assignmentContext):
        self.results.directive_values.append(ctx.getText())

    def enterCtl_action(self, ctx: SecLangParser.Ctl_actionContext):
        self.results.directive_values.append(ctx.getText())

    def enterCtl_id(self, ctx: SecLangParser.Ctl_idContext):
        self.results.directive_values.append(ctx.getText())

    def enterComment(self, ctx: SecLangParser.CommentContext):
        # ctx.COMMENT() can be None if there is only a HASH without comment text
        if ctx.COMMENT() is not None:
            self.results.comments.append(ctx.COMMENT().getText())
        else:
            self.results.comments.append("")

    def enterTransformation_action_value(self, ctx: SecLangParser.Transformation_action_valueContext):
        self.results.directive_values.append(ctx.getText())

    def enterComment_block(self, ctx: SecLangParser.Comment_blockContext):
        self.results.comment_blocks += 1